### PR TITLE
Issue #173 implement caching for aggregatedCostModel endpoint

### PR DESCRIPTION
**Pull request by nikovacevic**
Done
- Imported in [go-cache](https://github.com/patrickmn/go-cache) for TTL caching
- In `/aggregatedCostModel` handler, implemented parameter-keyed cache with default expiry of 2 minutes. Pass `clearCache=true` to clear the cache and `disableCache=true` to force a one-time cache miss.
- Added helper method for identifying cache hits and misses in HTTP responses
- Manual testing

Questions
- Should we attempt a second layer of caching of the `ComputeCostDataRange` results?
- This solution is tightly-coupled to the `AggregateCostModel` handler. Should we try to factor this out into a more general-purpose component, or save that for when we move caching out of the `cost-model`?
- Should we start by warming the cache with a call to `aggregate=namespace` and `window=1d` or save that for an improvement?

To Do
- Unit and/or integration tests
- Remove helper methods and TODO comments

### CODE REVIEW

**Comment by nikovacevic** [comment]
Tomorrow I'm planning to write integration tests for at least the following:
- two requests in rapid succession produce identical responses
- one request, followed by a `disableCache=true` request produce values within 0.1% of each other, but not necessarily identical
- one request, followed by a `clearCache=true` request produce values within 0.1% of each other, but not necessarily identical

It would be nice to make the expiration configurable so that we don't have to sleep for 2 minutes in the middle of the test to assert natural expiry, but something testing that is also on my radar.

File: main.go [hunk 1]
```diff
@@ -202,12 +230,25 @@ func (a *Accesses) ClusterCostsOverTime(w http.ResponseWriter, r *http.Request,
...
-	aggregation := r.URL.Query().Get("aggregation")
 	namespace := r.URL.Query().Get("namespace")
+	aggregationField := r.URL.Query().Get("aggregation")
 	aggregationSubField := r.URL.Query().Get("aggregationSubfield")
 
+	disableCache := r.URL.Query().Get("disableCache") == "true"
+	clearCache := r.URL.Query().Get("clearCache") == "true"
+
+	// TODO nikovacevic-caching this should be required, right? Can we return 400 Bad Request if not set?
+	if aggregationField == "" {
```
Comment (dwbrown2) [inline main.go]: Yep, seems fine to me. This is a required field. Nit: I may provide a more specific error message, maybe "Missing aggregationField parameter"?

File: main.go [hunk 2]
```diff
@@ -219,6 +260,8 @@ func (a *Accesses) AggregateCostModel(w http.ResponseWriter, r *http.Request, ps
 		endTime = endTime.Add(-1 * o)
 	}
 
+	// if window is defined in terms of days, convert it to hours
```
Comment (dwbrown2) [inline main.go]: thanks for these comments! :)

File: main.go [hunk 3]
```diff
@@ -832,6 +910,13 @@ func main() {
 	router.GET("/containerUptimes", a.ContainerUptimes)
 	router.GET("/aggregatedCostModel", a.AggregateCostModel)
 
+	router.GET("/image", func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		w.Header().Set("Content-Type", "application/json")
+		w.Header().Set("Access-Control-Allow-Origin", "*")
+		w.Write([]byte("nikovacevic/kubecost-cost-model:caching-4"))
```
Comment (dwbrown2) [inline main.go]: do we want to check this reference in?

File: go.mod [hunk 4]
```diff
@@ -39,3 +40,5 @@ require (
 	k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
 	k8s.io/klog v0.4.0
 )
+
+go 1.13
```
Comment (AjayTripathy) [inline go.mod]: we're still supporting 1.12 for now, what does this line do?

File: main.go (hunk 1 above)
Comment (AjayTripathy) [reply main.go]: SGTM

File: main.go [hunk 5]
```diff
@@ -239,23 +284,52 @@ func (a *Accesses) AggregateCostModel(w http.ResponseWriter, r *http.Request, ps
...
+
+	// TODO nikovacevic-caching cache raw cost data and unmarshal it
+	// rawKey := fmt.Sprintf("raw:%s:%s:%s:%s", start, end, "1h", namespace)
+	// data, found := a.Cache.Get(rawKey)
+	// if !found || disableCache ...
+
 	data, err := a.Model.ComputeCostDataRange(a.PrometheusClient, a.KubeClientSet, a.Cloud, start, end, "1h", namespace)
 	if err != nil {
-		w.Write(wrapData(nil, err))
+		w.Write(wrapDataWithMessage(nil, err, fmt.Sprintf("cache miss")))
```
Comment (AjayTripathy) [inline main.go]: could we see aggKey in the returned message?

**Comment by AjayTripathy** [comment]
Re: questions

* Should we attempt a second layer of caching of the ComputeCostDataRange results?
- Let's hold on this to see if this gets us most of the way to speeding things up.
*This solution is tightly-coupled to the AggregateCostModel handler. Should we try to factor this out into a more general-purpose component, or save that for when we move caching out of the cost-model?
- I think it's okay to tightly couple for now.
* Should we start by warming the cache with a call to aggregate=namespace and window=1d or save that for an improvement?
- Let's hold on this also for now until we think it's necessary to get the performance we need.

File: main.go (hunk 3 above)
Comment (nikovacevic) [reply main.go]: Nope, that whole endpoint is just for development purposes. Removing now.

File: go.mod (hunk 4 above)
Comment (nikovacevic) [reply go.mod]: That's the line my VS Code Go extension keeps adding. Until I find and disable the setting that controls it, I'll be careful to delete manually.

File: main.go (hunk 1 above)
Comment (nikovacevic) [reply main.go]: Agreed, changing now.

File: main.go (hunk 2 above)
Comment (nikovacevic) [reply main.go]: Sure thing!

**Comment by AjayTripathy** [comment]
@nikovacevic as discussed, had to get out #177 for a user issue before this, this LGTM after conflicts fixed.

**Pull request merged**